### PR TITLE
add AutoCastPass for precision sensitive operator

### PR DESCRIPTION
**Pull request by thisjiang**
为精度敏感型算子：

1. 数学类：`sin,cos,exp,log,log2,log10,sqrt,rsqrt,cbrt,erf,sinh,cosh,tanh,asin,acos,atan,asinh,acosh,atanh,sigmoid,remainder,pow`
2. 复杂算算子：`reduce_sum,reduce_prod,batch_norm_train,batch_norm_grad,softmax,gelu,sum`

添加`Cast`转换：输入`Cast`到fp32，使用fp32计算，输出`Cast`回fp16

### CODE REVIEW

**Review by Shixiaowei02 — APPROVED** [review]

**Comment by SunNy820828449** [comment]
bn的在fp16下的融合是不是会受到影响，因为现在没有recompute机制。

**Review by SunNy820828449 — APPROVED** [review]
LGTM

**Comment by thisjiang** [comment]
> bn的在fp16下的融合是不是会受到影响，因为现在没有recompute机制。

@SunNy820828449 我马上会提交另一个基于Graph的Pass：`CastRecomputePass`，若`cast`算子被多个算子用到，则每个算子都会创建一个单独的cast
